src/sentry/sentry_metrics/consumers/indexer/slicing_router.py

```diff
@@ -0,0 +1,83 @@
+from typing import MutableMapping, Optional
+
+from arroyo import Message, Topic
+from arroyo.backends.kafka import KafkaPayload
+from confluent_kafka import Producer
+from django.conf import settings
+
+from sentry.ingest.partitioning import (
+    Sliceable,
+    is_sliced,
+    map_logical_partition_to_slice,
+    map_org_id_to_logical_partition,
+)
+from sentry.sentry_metrics.consumers.indexer.routing_producer import MessageRoute, MessageRouter
+from sentry.utils import kafka_config
+
+
+class SlicingRouter(MessageRouter[KafkaPayload]):
+    def __init__(
+        self,
+        sliceable: Sliceable,
+        logical_output_topic: str,
+    ) -> None:
+        self.__sliceable = sliceable
+        self.__logical_output_topic = logical_output_topic
+        self.__slice_to_producer: MutableMapping[int, MessageRoute] = {}
+
+        if not is_sliced(self.__sliceable):
+            self.__slice_to_producer[0] = MessageRoute(
+                producer=Producer(
+                    kafka_config.get_kafka_producer_cluster_options(
+                        settings.KAFKA_TOPICS[self.__logical_output_topic]["cluster"]
+                    )
+                ),
+                topic=Topic(self.__logical_output_topic),
+            )
+            self.__slicing_enabled = False
+        else:
+            for (
+                current_sliceable,
+                current_slice_id,
+            ), config in settings.SLICED_KAFKA_BROKER_CONFIG.items():
+                self.__slice_to_producer[current_slice_id] = MessageRoute(
+                    producer=Producer(config),
+                    topic=Topic(
+                        settings.SLICED_KAFKA_TOPIC_MAP[(current_sliceable, current_slice_id)]
+                    ),
+                )
+            assert len(self.__slice_to_producer) == len(
+                settings.SENTRY_SLICING_CONFIG[sliceable].keys()
+            )
+            self.__slicing_enabled = True
+
+    def get_route_for_message(self, message: Message[KafkaPayload]) -> MessageRoute:
+        """
+        Get route for the message. If slicing is enabled, the message will be routed
+        based on the mapping of partition to slice. If slicing is disabled, the message
+        will be routed to the default topic.
+        """
+        if not self.__slicing_enabled:
+            return self.__slice_to_producer[0]
+
+        org_id: Optional[int] = None
+        for header in message.payload.headers:
+            if header[0] == "org_id":
```

Comment:
Who's going to be setting this?

Reply:
This gets set from the preceding step of the processing strategy. Here's how it would be set https://github.com/getsentry/sentry/pull/40867/files#diff-7697ef4fd4cf439ed7ad41b7d98c00ff7d24cde30953e12d6bcb877303bb942dR350.

The payload is already serialized into JSON string at the stage. And I was avoiding deserializing again over here for performance reasons. Header based routing seemed like the most optimal way to perform routing